### PR TITLE
Allow other buffer types to be passed to populateMatrix

### LINKED ISSUE #20: Allow other buffer types to be passed to populateMatrix
The API should not restrict passed data type accepted by populateMatrix to Float32Array.

Proposed idl changes.

```
interface OrientationSensor : Sensor {
  ...
  void populateMatrix(Float32Array or Float64Array or DOMMatrix targetMatrix);
};
```

### LINKED ISSUE #24: Define populateMatrix algorithm taking into account projection / translation values
If the matrix provided to populateMatrix already contains translation / scale / projection data, specification should define algorithm that will cover such cases, e.g. 

- overwrite everything
- check if is projection => overwrite
- check if translation => keep translation, add rotation

### PR DESCRIPTION
**Pull request by alexshalamov**
Fixes #20
Fixes #24

***
[Preview](https://s3.amazonaws.com/pr-preview/alexshalamov/orientation-sensor/gh-pages.html) | [Diff](https://s3.amazonaws.com/pr-preview/w3c/orientation-sensor/b842f61...alexshalamov:8ba11fa.html)

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -214,45 +215,56 @@ In other words, this attribute returns [=latest reading=]["quaternion"].
...
- - X = V<sub>x</sub> * sin(θ/2).
- - Y = V<sub>y</sub> * sin(θ/2).
- - Z = V<sub>z</sub> * sin(θ/2).
+ - W = cos(θ/2)
+ - X = V<sub>x</sub> * sin(θ/2)
+ - Y = V<sub>y</sub> * sin(θ/2)
+ - Z = V<sub>z</sub> * sin(θ/2)
 
-The rotation matrix is flattened in |targetBuffer| according to the column-major order.
+The rotation matrix is flattened in |targetMatrix| according to the column-major order.
```
Comment (pozdnyakov) [inline index.bs]: can it apply to DOMMatrix?

Comment (alexshalamov) [inline index.bs]: Yes, rotation matrix is calculated from quat as shown in the illustration and flattened into array or m11-m44 fields. So, I think this sentence is valid.

**Comment by alexshalamov** [comment]
@kenchris PTAL

File: index.bs [hunk 2]
```diff
@@ -214,45 +215,68 @@ In other words, this attribute returns [=latest reading=]["quaternion"].
...
 
 <img src="quaternion_to_rotation_matrix.png" style="display: block;margin: auto;" alt="Converting quaternion to rotation matrix." width="50%" height="50%">
 
 where:
 
- - W = cos(θ/2).
- - X = V<sub>x</sub> * sin(θ/2).
- - Y = V<sub>y</sub> * sin(θ/2).
- - Z = V<sub>z</sub> * sin(θ/2).
+ - W = cos(θ/2)
```
Comment (kenchris) [inline index.bs]: Cant we make these lower case? That seems more common (yes I know that upper case is often used on graphs for the axises )

File: index.bs (hunk 1 above)
Comment (kenchris) [inline index.bs]: Should you add a note about that then?

File: index.bs (hunk 2 above)
Comment (alexshalamov) [inline index.bs]: done

File: index.bs (hunk 1 above)
Comment (alexshalamov) [inline index.bs]: done

**Comment by kenchris** [comment]
Lgtm

File: index.bs (hunk 2 above)
Comment (pozdnyakov) [inline index.bs]: this is not in sync with the figure above (with the matrix)

Comment (pozdnyakov) [inline index.bs]: @kenchris shall we modify matrix elements so that they are low case too?

Comment (alexshalamov) [inline index.bs]: @pozdnyakov true, in the reference [QUATERNIONS], upper case is used, so I think I have to revert back to caps. @kenchris wdyt?

**Comment by kenchris** [comment]
Hmm not sure, that is not referring to a real spec right. @anssik

**Comment by alexshalamov** [comment]
@kenchris you mean TR link?

**Comment by anssiko** [comment]
@kenchris, can you define "that" in "that is not referring to a real spec right"?

**Comment by kenchris** [comment]
I was referring to http://www.emis.ams.org/proceedings/Varna/vol1/GEOM09.pdf

Actually from quickly looking over that PDF, it seems to define the uppercase versions as vector sets, and not as individual vectors. That makes sense, in Math, sets/groups are usually uppercase (or special version of upper case letters)

**Comment by anssiko** [comment]
>[QUATERNIONS]
Kuipers, Jack B. Quaternions and rotation sequences. Vol. 66.. 1999. Informational. URL: http://www.emis.ams.org/proceedings/Varna/vol1/GEOM09.pdf

... referenced in:

> The elements of the list are equal to components of a unit quaternion [QUATERNIONS]

... seems like a legit informative reference.

**Comment by alexshalamov** [comment]
> I was referring to http://www.emis.ams.org/proceedings/Varna/vol1/GEOM09.pdf

We have two references, one for unit quaternion [QUATERNIONS] and one for converting quaternion to rotation matrix [QUATCONV]

File: index.bs (hunk 2 above)
Comment (alexshalamov) [inline index.bs]: correction [QUATERNIONS] => [QUATCONV]

**Comment by kenchris** [comment]
https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion
https://en.wikipedia.org/wiki/Rotation_formalisms_in_three_dimensions#Conversion_formulae_between_formalisms

**Comment by alexshalamov** [comment]
@kenchris I think we should use proper references, published books, papers. Links to wikipedia is not a good reference. Entry can be modified, deleted, etc. @anssiko

**Comment by kenchris** [comment]
I am sure we must be able to find another reference... it looks like basic knowledge.. all references are to conversion from the other aside (rot -> quat)

**Comment by alexshalamov** [comment]
@kenchris Created Issue #25, this PR doesn't add, nor remove any references, so I think we can land it and handle upper case / lower case / references issue separately. wdyt?

**Comment by kenchris** [comment]
https://d3cw3dd2w32x2b.cloudfront.net/wp-content/uploads/2015/01/matrix-to-quat.pdf insomniac games :)

**Comment by alexshalamov** [comment]
@kenchris Insomniac games pdf hosted somewhere on wordpress :suspect: let's it proceed with this PR and fix references for #25 later :D

**Comment by pozdnyakov** [comment]
yeah, let's proceed this discussion in #25

**Pull request merged**